api/src/sparql.rs

```diff
@@ -16,19 +16,101 @@
 
 use crate::term::TTerm;
 use crate::triple::stream::TripleSource;
+use std::borrow::Borrow;
 use std::error::Error;
 
 /// A dataset that can be queried with SPARQL.
 pub trait SparqlDataset {
+    /// The type of terms that SELECT queries will return.
     type BindingsTerm: TTerm;
+    /// The type of bindings that SELECT queries will return.
     type BindingsResult: SparqlBindings<Self>;
+    /// The type of triples that GRAPH and DESCRIBE queries will return.
     type TriplesResult: TripleSource;
+    /// The type of errors that processing SPARQL queries may raise.
     type SparqlError: Error + 'static;
+    /// The type representing pre-processed queries.
+    ///
+    /// See [`prepare_query`](#tymethod.prepare_query) for more defail.
+    type Query: Query<Error=Self::SparqlError>;
+
+    /// Parse and immediately execute `query`.
+    ///
+    /// `query` is usually either a `&str` that will be parsed on the fly,
+    /// or a `Self::Query` that was earlier prepared by the [`prepare_query`] method.
+    ///
+    /// [`prepare_query`]: #method.prepared
+    fn query<Q>(&self, query: Q) -> Result<SparqlResult<Self>, Self::SparqlError>
+    where Q: ToQuery<Self::Query>;
+
+    /// Prepare a query for multiple future executions.
+    ///
+    /// This allows some implementation to mutualize parsing,
```

Comment:
I'm not sure I understand what you mean by "mutualize".

Reply:
What I meant was "parse once, execute many times", hence the idea of "mutualizing".
But I agree that this is not the best choice of word. I'll change that as well.